script.program.homeassistant/addon.py

```diff
@@ -0,0 +1,463 @@
+import os
+import sys
+import urllib
+import urllib.parse
+
+import xbmcgui
+import xbmcplugin
+import xbmcaddon
+from xbmc import log as xbmc_log
+
+import requests
+import yaml
+import calendar
+import time
+from urllib.parse import parse_qsl
+from datetime import datetime
+import iso8601
+
+base_url = sys.argv[0]
+addon_handle = int(sys.argv[1])
+args = urllib.parse.parse_qs(sys.argv[2][1:])
+
+__addon__ = xbmcaddon.Addon()
+__addonname__ = __addon__.getAddonInfo('name')
+__icon__ = __addon__.getAddonInfo('icon')
+__addonid__ = __addon__.getAddonInfo('id')
+
+mode = args.get('mode', None)
+
+imgIconResourcePath = os.path.join(xbmcaddon.Addon().getAddonInfo('path'),'resources','img','icon')
+imgFanartResourcePath = os.path.join(xbmcaddon.Addon().getAddonInfo('path'),'resources','img','fanart')
+fileHakaFavourites = os.path.join(xbmcaddon.Addon().getAddonInfo('path'),'hakaFavourites.yaml')
+dummyWav = os.path.join(xbmcaddon.Addon().getAddonInfo('path'),'resources','1.wav')
+
+haDomainNames = 		['automation','climate','group','light','scene','script','sensor','switch','vacuum','person']
+haDomainSettings = 		[ False, False, False, False, False, False, False, False, False, False, False]
+haDomainTranslations = 	[30005,30006,30007,30008,30009,30010,30011,30012,30013,30014,30019]
+
+haServer = __addon__.getSetting('haServer')
+haToken = __addon__.getSetting('haToken')
+
+api_base = haServer + '/api'
+headers = {'Authorization': 'Bearer ' + haToken, 'Content-Type': 'application/json'}
+
+def build_url(query):
+	return base_url + '?' + urllib.parse.urlencode(query)
+
+def utc_to_local(utc_dt):
+	# get integer timestamp to avoid precision lost
+	timestamp = calendar.timegm(utc_dt.timetuple())
+	local_dt = datetime.fromtimestamp(timestamp)
+	assert utc_dt.resolution >= timedelta(microseconds=1)
+	return local_dt.replace(microsecond=utc_dt.microsecond)
+
+def parse_date(event_date):
+	datetime_obj = iso8601.parse_date(event_date)
+	datetime_obj = utc_to_local(datetime_obj)
+	dateString = datetime_obj.strftime(xbmc.getRegion('dateshort'))
+	return dateString
+
+def parse_time(event_date):
+	datetime_obj = iso8601.parse_date(event_date)
+	datetime_obj = utc_to_local(datetime_obj)
+	timeString = datetime_obj.strftime(xbmc.getRegion('time'))
+	return timeString
+	
+def parse_dateTime(event_date):
+	return parse_date(event_date) + ' ' + parse_time(event_date)
+
+def have_credentials():
+	return haServer and haToken
+
+def show_dialog(message):
+	xbmcgui.Dialog().ok(__addonname__, message)
+	
+def log(txt, loglevel=xbmc.LOGWARNING):
+	if __addon__.getSetting( "logEnabled" ) == "true":
+		message = u'%s: %s' % (__addonid__, txt)
+		xbmc.log(msg=message, level=loglevel)
+
+def read_favourites():
+	favourites = []
+	if os.stat(fileHakaFavourites).st_size > 0:
+		try:
+			file = open(fileHakaFavourites, 'r')
```

Comment:
please use python's with statement:

```python
with open(fileHakaFavourites, 'r') as f:
   favourites = yaml.load(f, Loader=yaml.BaseLoader)
```

This will ensure the filehandler is always closed